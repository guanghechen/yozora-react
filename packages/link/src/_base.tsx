import styled from 'styled-components'
import { defaultLinkTheme, getLinkStyle } from './theme'


/**
 * Props for creating link
 */
export interface LinkProps extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
  /**
   * Link url
   */
  url: string
  /**
   * Link title
   */
  title?: string
  /**
   * link content
   */
  children: React.ReactNode
}


export const Container = styled.a`
  color: ${ getLinkStyle('color') };
  font-size: ${ getLinkStyle('fontSize') };
  font-style: ${ getLinkStyle('fontStyle') };
  text-decoration: ${ getLinkStyle('textDecoration') };
`


Container.defaultProps = {
  theme: { yozora: { link: defaultLinkTheme } }
}
