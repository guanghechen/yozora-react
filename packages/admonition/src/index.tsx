import cn from 'clsx'
import PropTypes from 'prop-types'
import React from 'react'

export interface AdmonitionProps {
  /**
   * Admonition keyword.
   */
  keyword?: string
  /**
   * Admonition title
   */
  title?: React.ReactNode
  /**
   * Admonition contents.
   */
  children?: React.ReactNode
  /**
   * Root css class of the component.
   */
  className?: string
  /**
   * Root css style.
   */
  style?: React.CSSProperties
}

/**
 * Render yozora `admonition`.
 *
 * @see https://www.npmjs.com/package/@yozora/ast#admonition
 * @see https://www.npmjs.com/package/@yozora/tokenizer-admonition
 */
export function Admonition(props: AdmonitionProps): React.ReactElement {
  const { className, style, keyword = 'default', title, children } = props

  let modifier = keyword.trim().toLowerCase()
  switch (modifier) {
    case 'note':
      modifier = 'default'
      break
    case 'important':
      modifier = 'info'
      break
    case 'tip':
      modifier = 'success'
      break
    case 'caution':
      modifier = 'warning'
      break
    case 'error':
      modifier = 'danger'
      break
  }

  return (
    <div
      className={cn(
        'yozora-admonition',
        'yozora-admonition--' + modifier,
        className,
      )}
      style={style}
    >
      <div key="heading" className="yozora-admonition__heading">
        <h5>{title}</h5>
      </div>
      <div key="body" className="yozora-admonition__body">
        {children}
      </div>
    </div>
  )
}

Admonition.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  keyword: PropTypes.string,
  style: PropTypes.object,
  title: PropTypes.node,
}

Admonition.displayName = 'YozoraAdmonition'
export default Admonition
