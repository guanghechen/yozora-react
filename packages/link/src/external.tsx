import React from 'react'
import PropTypes from 'prop-types'
import { Container, LinkProps } from './_base'


/**
 * Render `link` content
 *
 * @param props
 */
export const ExternalLink = React.forwardRef<HTMLAnchorElement, LinkProps>(
  (props, forwardRef): React.ReactElement => {
    const {
      url,
      href,
      title,
      target = '_blank',
      rel = 'noreferrer',
      ...htmlProps
    } = props

    return (
      <Container
        { ...htmlProps }
        ref={ forwardRef }
        title={ title }
        href={ url }
        target={ target }
        rel={ rel }
      />
    )
  }
)


ExternalLink.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string,
  children: PropTypes.node.isRequired,
}


ExternalLink.displayName = 'YozoraExternalLink'
export default ExternalLink
