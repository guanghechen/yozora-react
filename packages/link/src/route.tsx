import PropTypes from 'prop-types'
import React from 'react'
import { Link } from 'react-router-dom'
import type { LinkProps } from './_base'
import { Container } from './_base'

/**
 * Render `link` content
 *
 * @param props
 */
export const RouteLink = React.forwardRef<HTMLAnchorElement, LinkProps>(
  (props, forwardRef): React.ReactElement => {
    const { url, href, title, ...htmlProps } = props

    return (
      <Container
        {...htmlProps}
        as={Link}
        ref={forwardRef}
        title={title}
        to={{ pathname: url }}
      />
    )
  },
)

RouteLink.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string,
  children: PropTypes.node.isRequired,
  href: PropTypes.string,
}

RouteLink.displayName = 'YozoraRouteLink'
export default RouteLink
