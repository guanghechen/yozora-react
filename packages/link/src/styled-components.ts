import type { YozoraLinkTheme } from './theme'
import 'styled-components'


declare module 'styled-components' {
  export interface DefaultTheme {
    yozora: YozoraTheme
  }

  interface YozoraTheme {
    link?: YozoraLinkTheme
  }
}
