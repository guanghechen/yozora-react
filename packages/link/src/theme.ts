import type { CSSProperties, DefaultTheme } from 'styled-components'


/**
 * yozora `link` theme
 */
export interface YozoraLinkTheme {
  /**
   * Link font color
   * @default 'inherit'
   */
  color?: CSSProperties['color'] | string
  /**
   * Link font size
   * @default 'inherit'
   */
  fontSize?: CSSProperties['fontSize'] | string
  /**
   * Link font style
   * @default 'inherit'
   */
  fontStyle?: CSSProperties['fontStyle'] | string
  /**
   * Link text decoration
   * @default 'none'
   */
  textDecoration?: CSSProperties['textDecoration'] | string
}


/**
 * Default yozora `link` theme
 */
export const defaultLinkTheme: YozoraLinkTheme = {
  color: 'inherit',
  fontSize: 'inherit',
  fontStyle: 'inherit',
  textDecoration: 'none',
}


/**
 * Get `link` style
 * @param key
 * @param defaultTheme
 */
export function getLinkStyle(
  key: keyof YozoraLinkTheme,
  defaultTheme: YozoraLinkTheme = defaultLinkTheme,
) {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  return (props: { theme: DefaultTheme }) => {
    const { yozora } = props.theme
    if (
      yozora == null ||
      yozora.link == null ||
      yozora.link[key] == null
    ) return defaultTheme[key]
    return yozora.link[key]
  }
}
