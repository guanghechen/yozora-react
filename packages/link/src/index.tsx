import PropTypes from 'prop-types'
import React from 'react'
import type { LinkProps } from './_base'
import { Container } from './_base'
import ExternalLink from './external'
import RouteLink from './route'
import './styled-components'

export { ExternalLink } from './external'
export { RouteLink } from './route'
export * from './theme'

/**
 * Test if the given url is an external link
 *
 * @param url
 * @see https://github.com/mrded/is-url-external/blob/f46b37811034048b0803966af7832f3c7822da5a/index.js
 */
export function isExternalUrl(url: string): boolean {
  const host = window.location.hostname

  // Absolute URL.
  if (/^https?:\/\//.test(url)) {
    // The easy way to parse an URL, is to create <a> element.
    // @see: https://gist.github.com/jlong/2428561
    const anchor = document.createElement('a')
    anchor.href = url
    return host !== anchor.hostname
  }

  return false
}

/**
 * Render Link depend on whether if it's a external link
 */
export const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(
  (props, forwardRef): React.ReactElement => {
    const isExternal = isExternalUrl(props.url)
    return isExternal ? (
      <ExternalLink {...props} ref={forwardRef} />
    ) : (
      <RouteLink {...props} ref={forwardRef} />
    )
  },
)

Link.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string,
  children: PropTypes.node,
}

Link.displayName = 'YozoraLink'
export default Link

export const LinkClasses = {
  container: `${Container}`,
}
