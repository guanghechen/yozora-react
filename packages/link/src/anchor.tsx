import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { defaultLinkTheme, getLinkStyle } from './theme'


/**
 * Props for creating link
 */
export interface LinkProps extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
  /**
   * Link url
   */
  url: string
  /**
   * Link title
   */
  title?: string
  /**
   * link content
   */
  children: React.ReactNode
}


/**
 * Render `link` content
 *
 * @param props
 */
export const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(
  (props, forwardRef): React.ReactElement => {
    const {
      url,
      href,
      title,
      ...htmlProps
    } = props

    return (
      <Container
        { ...htmlProps }
        ref={ forwardRef }
        title={ title }
        href={ url }
      />
    )
  }
)


Link.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string,
  children: PropTypes.node.isRequired,
}


Link.displayName = 'YozoraLink'
export default Link


export const Container = styled.a`
  color: ${ getLinkStyle('color') };
  font-size: ${ getLinkStyle('fontSize') };
  font-style: ${ getLinkStyle('fontStyle') };
  text-decoration: ${ getLinkStyle('textDecoration') };
`


Container.defaultProps = {
  theme: { yozora: { link: defaultLinkTheme } }
}


export const LinkClasses = {
  container: `${ Container }`,
}
