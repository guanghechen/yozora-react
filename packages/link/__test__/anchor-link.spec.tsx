import React from 'react'
import { mount, render } from 'enzyme'
import { DefaultTheme, ThemeProvider } from 'styled-components'
import { ExternalLink as Link } from '../src'


describe('basic rendering case', () => {
  const errorLogger = jest
    .spyOn(global.console, 'error')
    .mockImplementation((...args) => {
      throw new Error(args.join(' '))
    })

  afterAll(() => {
    errorLogger.mockRestore()
  })

  it('render a simple content', () => {
    const text = 'Hello, world!'
    const wrapper = render(
      <Link url="#">{ text }</Link>
    )
    expect(wrapper.text()).toEqual(text)
  })

  it('render with custom className', () => {
    const text = 'Hello, world!'
    const className = 'custom-link'
    const wrapper = render(
      <Link url="#" className={ className }>
        <span>{ text }</span>
      </Link>
    )
    expect(wrapper.hasClass(className)).toEqual(true)
    expect(wrapper.text()).toEqual(text)
  })

  it('url and children both are required', () => {
    for (const value of [undefined, null] as any[]) {
      expect(() => {
        render(<Link url={ value }>link text</Link>
        )
      }).toThrow(/Failed prop type: The prop `url` is marked as required/i)

      expect(() => {
        render(<Link url="/home">{ value }</Link>)
      }).toThrow(/Failed prop type: The prop `children` is marked as required/i)
    }
  })

  it('forward ref', () => {
    const ref = React.createRef<HTMLAnchorElement>()
    const wrapper = mount(
      <Link
        ref={ ref }
        url="/home"
        data-value="waw"
      >
        1
      </Link>
    )

    const o = wrapper.getDOMNode()
    expect(o).toEqual(ref.current)
    expect(o.getAttribute('data-value')).toEqual('waw')
  })

  it('snapshot', () => {
    const wrapper = render(
      <Link
        url="/home"
        title="home"
        style={{ color: 'orange', fontSize: '16px' }}
      >
        some text1
        <span>some text2</span>
      </Link>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('snapshot with theme', () => {
    const theme: DefaultTheme = {
      yozora: {
        link: {
          color: 'blue',
          fontSize: undefined,
          fontStyle: 'italic',
          textDecoration: 'none',
        }
      }
    }

    const wrapper = mount(
      <ThemeProvider theme={ theme }>
        <Link
          url="/home"
          title="home"
          style={{ color: 'orange', fontSize: '16px' }}
        >
          some text1
          <span>some text2</span>
        </Link>
      </ThemeProvider>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
