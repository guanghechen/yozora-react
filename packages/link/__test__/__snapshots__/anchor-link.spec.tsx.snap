// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic rendering case snapshot 1`] = `
.c0 {
  color: inherit;
  font-size: inherit;
  font-style: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<a
  class="c0"
  href="/home"
  rel="noreferrer"
  style="color:orange;font-size:16px"
  target="_blank"
  title="home"
>
  some text1
  <span>
    some text2
  </span>
</a>
`;

exports[`basic rendering case snapshot with theme 1`] = `
.c0 {
  color: blue;
  font-size: inherit;
  font-style: italic;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Component
  theme={
    Object {
      "yozora": Object {
        "link": Object {
          "color": "blue",
          "fontSize": undefined,
          "fontStyle": "italic",
          "textDecoration": "none",
        },
      },
    }
  }
>
  <YozoraExternalLink
    style={
      Object {
        "color": "orange",
        "fontSize": "16px",
      }
    }
    title="home"
    url="/home"
  >
    <styled.a
      href="/home"
      rel="noreferrer"
      style={
        Object {
          "color": "orange",
          "fontSize": "16px",
        }
      }
      target="_blank"
      theme={
        Object {
          "yozora": Object {
            "link": Object {
              "color": "inherit",
              "fontSize": "inherit",
              "fontStyle": "inherit",
              "textDecoration": "none",
            },
          },
        }
      }
      title="home"
    >
      <a
        className="c0"
        href="/home"
        rel="noreferrer"
        style={
          Object {
            "color": "orange",
            "fontSize": "16px",
          }
        }
        target="_blank"
        title="home"
      >
        some text1
        <span>
          some text2
        </span>
      </a>
    </styled.a>
  </YozoraExternalLink>
</Component>
`;
